// jimp/es/constants.js
// used to auto resizing etc.
export var AUTO = -1; // supported mime types

export var MIME_PNG = 'image/png';
export var MIME_TIFF = 'image/tiff';
export var MIME_JPEG = 'image/jpeg';
export var MIME_JGD = 'image/jgd';
export var MIME_BMP = 'image/bmp';
export var MIME_X_MS_BMP = 'image/x-ms-bmp';
export var MIME_GIF = 'image/gif'; // PNG filter types

export var PNG_FILTER_AUTO = -1;
export var PNG_FILTER_NONE = 0;
export var PNG_FILTER_SUB = 1;
export var PNG_FILTER_UP = 2;
export var PNG_FILTER_AVERAGE = 3;
export var PNG_FILTER_PATH = 4;
export var RESIZE_NEAREST_NEIGHBOR = 'nearestNeighbor';
export var RESIZE_BILINEAR = 'bilinearInterpolation';
export var RESIZE_BICUBIC = 'bicubicInterpolation';
export var RESIZE_HERMITE = 'hermiteInterpolation';
export var RESIZE_BEZIER = 'bezierInterpolation'; // Align modes for cover, contain, bit masks

export var HORIZONTAL_ALIGN_LEFT = 1;
export var HORIZONTAL_ALIGN_CENTER = 2;
export var HORIZONTAL_ALIGN_RIGHT = 4;
export var VERTICAL_ALIGN_TOP = 8;
export var VERTICAL_ALIGN_MIDDLE = 16;
export var VERTICAL_ALIGN_BOTTOM = 32; // blend modes

export var BLEND_SOURCE_OVER = 'srcOver';
export var BLEND_DESTINATION_OVER = 'dstOver';
export var BLEND_MULTIPLY = 'multiply';
export var BLEND_SCREEN = 'screen';
export var BLEND_OVERLAY = 'overlay';
export var BLEND_DARKEN = 'darken';
export var BLEND_LIGHTEN = 'lighten';
export var BLEND_HARDLIGHT = 'hardLight';
export var BLEND_DIFFERENCE = 'difference';
export var BLEND_EXCLUSION = 'exclusion'; // Discover Jimp directory in any environment. (also in fucking karma)

function getJimpDir() {
  if ("module" === 'test' && undefined === 'browser') {
    return 'http://localhost:9876/base/test/';
  }

  return "".concat(__dirname, "/");
}

export var dirName = getJimpDir(); // Font locations

export var FONT_SANS_8_BLACK = dirName + '../fonts/open-sans/open-sans-8-black/open-sans-8-black.fnt';
export var FONT_SANS_10_BLACK = dirName + '../fonts/open-sans/open-sans-10-black/open-sans-10-black.fnt';
export var FONT_SANS_12_BLACK = dirName + '../fonts/open-sans/open-sans-12-black/open-sans-12-black.fnt';
export var FONT_SANS_14_BLACK = dirName + '../fonts/open-sans/open-sans-14-black/open-sans-14-black.fnt';
export var FONT_SANS_16_BLACK = dirName + '../fonts/open-sans/open-sans-16-black/open-sans-16-black.fnt';
export var FONT_SANS_32_BLACK = dirName + '../fonts/open-sans/open-sans-32-black/open-sans-32-black.fnt';
export var FONT_SANS_64_BLACK = dirName + '../fonts/open-sans/open-sans-64-black/open-sans-64-black.fnt';
export var FONT_SANS_128_BLACK = dirName + '../fonts/open-sans/open-sans-128-black/open-sans-128-black.fnt';
export var FONT_SANS_8_WHITE = dirName + '../fonts/open-sans/open-sans-8-white/open-sans-8-white.fnt';
export var FONT_SANS_16_WHITE = dirName + '../fonts/open-sans/open-sans-16-white/open-sans-16-white.fnt';
export var FONT_SANS_32_WHITE = dirName + '../fonts/open-sans/open-sans-32-white/open-sans-32-white.fnt';
export var FONT_SANS_64_WHITE = dirName + '../fonts/open-sans/open-sans-64-white/open-sans-64-white.fnt';
export var FONT_SANS_128_WHITE = dirName + '../fonts/open-sans/open-sans-128-white/open-sans-128-white.fnt'; // Edge Handling

export var EDGE_EXTEND = 1;
export var EDGE_WRAP = 2;
export var EDGE_CROP = 3;
