// @bochilteam/scraper-news/lib/cjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.merdeka = exports.liputan6 = exports.suaracom = exports.kompas = exports.antaranews = exports.cnbcindonesia = void 0;
const cnbcindonesia_js_1 = __importDefault(require("./src/cnbcindonesia.js"));
exports.cnbcindonesia = cnbcindonesia_js_1.default;
const antaranews_js_1 = __importDefault(require("./src/antaranews.js"));
exports.antaranews = antaranews_js_1.default;
const kompas_js_1 = __importDefault(require("./src/kompas.js"));
exports.kompas = kompas_js_1.default;
const suaracom_js_1 = __importDefault(require("./src/suaracom.js"));
exports.suaracom = suaracom_js_1.default;
const liputan6_js_1 = __importDefault(require("./src/liputan6.js"));
exports.liputan6 = liputan6_js_1.default;
const merdeka_js_1 = __importDefault(require("./src/merdeka.js"));
exports.merdeka = merdeka_js_1.default;
