// dhn-api/index.js
const news = require("./database/news.js");
const anime = require("./database/manga.js");
const search = require("./database/search.js");
const other = require("./database/other.js");
const json = require("./database/json.js");

exports.BBCNews = news.BBC;
exports.metroNews = news.metroTV_;
exports.CNNNews = news.CNN_;
exports.iNews = news.iNewsTV_;
exports.KumparanNews = news.Kumparan_;
exports.TribunNews = news.Tribun_;
exports.DailyNews = news.DailyNews_;
exports.DetikNews = news.DetikNews_;
exports.OkezoneNews = news.Okezone_;
exports.CNBCNews = news.CNBC_;
exports.FajarNews = news.KoranFajar_;
exports.KompasNews = news.Kompas_;
exports.SindoNews = news.KoranSindo_;
exports.TempoNews = news.TempoNews_;
exports.IndozoneNews = news.Indozone_;
exports.AntaraNews = news.AntaraNews_;
exports.RepublikaNews = news.Republika_;
exports.VivaNews = news.VIVA_;
exports.KontanNews = news.Kontan_;
exports.MerdekaNews = news.Merdeka_;
exports.KomikuSearch = anime.Komiku_;
exports.AniPlanetSearch = anime.AnimePlanet_;
exports.KomikFoxSearch = anime.KomikFox_;
exports.KomikStationSearch = anime.KomikStation_;
exports.MangakuSearch = anime.Mangakus_;
exports.KiryuuSearch = anime.Kiryuus_;
exports.KissMangaSearch = anime.KissM_;
exports.KlikMangaSearch = anime.KlikS_;
exports.PalingMurah = search.palingmurah_;
exports.LayarKaca21 = search.layarkaca_;
exports.AminoApps = search.Amino_;
exports.Mangatoon = search.Mangatoon;
exports.WAModsSearch = search.WAMods;
exports.Emojis = other.Emojing_;
exports.CoronaInfo = other.Corona_;
exports.Cerpen = json.Cerpen_;
exports.Quotes = json.Quotes_;
exports.Couples = json.Couples;
exports.JalanTikusMeme = json.JTMimers;
exports.Darkjokes = json.Dark;
