// jimp/dist/constants.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EDGE_CROP = exports.EDGE_WRAP = exports.EDGE_EXTEND = exports.FONT_SANS_128_WHITE = exports.FONT_SANS_64_WHITE = exports.FONT_SANS_32_WHITE = exports.FONT_SANS_16_WHITE = exports.FONT_SANS_8_WHITE = exports.FONT_SANS_128_BLACK = exports.FONT_SANS_64_BLACK = exports.FONT_SANS_32_BLACK = exports.FONT_SANS_16_BLACK = exports.FONT_SANS_14_BLACK = exports.FONT_SANS_12_BLACK = exports.FONT_SANS_10_BLACK = exports.FONT_SANS_8_BLACK = exports.dirName = exports.BLEND_EXCLUSION = exports.BLEND_DIFFERENCE = exports.BLEND_HARDLIGHT = exports.BLEND_LIGHTEN = exports.BLEND_DARKEN = exports.BLEND_OVERLAY = exports.BLEND_SCREEN = exports.BLEND_MULTIPLY = exports.BLEND_DESTINATION_OVER = exports.BLEND_SOURCE_OVER = exports.VERTICAL_ALIGN_BOTTOM = exports.VERTICAL_ALIGN_MIDDLE = exports.VERTICAL_ALIGN_TOP = exports.HORIZONTAL_ALIGN_RIGHT = exports.HORIZONTAL_ALIGN_CENTER = exports.HORIZONTAL_ALIGN_LEFT = exports.RESIZE_BEZIER = exports.RESIZE_HERMITE = exports.RESIZE_BICUBIC = exports.RESIZE_BILINEAR = exports.RESIZE_NEAREST_NEIGHBOR = exports.PNG_FILTER_PATH = exports.PNG_FILTER_AVERAGE = exports.PNG_FILTER_UP = exports.PNG_FILTER_SUB = exports.PNG_FILTER_NONE = exports.PNG_FILTER_AUTO = exports.MIME_GIF = exports.MIME_X_MS_BMP = exports.MIME_BMP = exports.MIME_JGD = exports.MIME_JPEG = exports.MIME_TIFF = exports.MIME_PNG = exports.AUTO = void 0;
// used to auto resizing etc.
var AUTO = -1; // supported mime types

exports.AUTO = AUTO;
var MIME_PNG = 'image/png';
exports.MIME_PNG = MIME_PNG;
var MIME_TIFF = 'image/tiff';
exports.MIME_TIFF = MIME_TIFF;
var MIME_JPEG = 'image/jpeg';
exports.MIME_JPEG = MIME_JPEG;
var MIME_JGD = 'image/jgd';
exports.MIME_JGD = MIME_JGD;
var MIME_BMP = 'image/bmp';
exports.MIME_BMP = MIME_BMP;
var MIME_X_MS_BMP = 'image/x-ms-bmp';
exports.MIME_X_MS_BMP = MIME_X_MS_BMP;
var MIME_GIF = 'image/gif'; // PNG filter types

exports.MIME_GIF = MIME_GIF;
var PNG_FILTER_AUTO = -1;
exports.PNG_FILTER_AUTO = PNG_FILTER_AUTO;
var PNG_FILTER_NONE = 0;
exports.PNG_FILTER_NONE = PNG_FILTER_NONE;
var PNG_FILTER_SUB = 1;
exports.PNG_FILTER_SUB = PNG_FILTER_SUB;
var PNG_FILTER_UP = 2;
exports.PNG_FILTER_UP = PNG_FILTER_UP;
var PNG_FILTER_AVERAGE = 3;
exports.PNG_FILTER_AVERAGE = PNG_FILTER_AVERAGE;
var PNG_FILTER_PATH = 4;
exports.PNG_FILTER_PATH = PNG_FILTER_PATH;
var RESIZE_NEAREST_NEIGHBOR = 'nearestNeighbor';
exports.RESIZE_NEAREST_NEIGHBOR = RESIZE_NEAREST_NEIGHBOR;
var RESIZE_BILINEAR = 'bilinearInterpolation';
exports.RESIZE_BILINEAR = RESIZE_BILINEAR;
var RESIZE_BICUBIC = 'bicubicInterpolation';
exports.RESIZE_BICUBIC = RESIZE_BICUBIC;
var RESIZE_HERMITE = 'hermiteInterpolation';
exports.RESIZE_HERMITE = RESIZE_HERMITE;
var RESIZE_BEZIER = 'bezierInterpolation'; // Align modes for cover, contain, bit masks

exports.RESIZE_BEZIER = RESIZE_BEZIER;
var HORIZONTAL_ALIGN_LEFT = 1;
exports.HORIZONTAL_ALIGN_LEFT = HORIZONTAL_ALIGN_LEFT;
var HORIZONTAL_ALIGN_CENTER = 2;
exports.HORIZONTAL_ALIGN_CENTER = HORIZONTAL_ALIGN_CENTER;
var HORIZONTAL_ALIGN_RIGHT = 4;
exports.HORIZONTAL_ALIGN_RIGHT = HORIZONTAL_ALIGN_RIGHT;
var VERTICAL_ALIGN_TOP = 8;
exports.VERTICAL_ALIGN_TOP = VERTICAL_ALIGN_TOP;
var VERTICAL_ALIGN_MIDDLE = 16;
exports.VERTICAL_ALIGN_MIDDLE = VERTICAL_ALIGN_MIDDLE;
var VERTICAL_ALIGN_BOTTOM = 32; // blend modes

exports.VERTICAL_ALIGN_BOTTOM = VERTICAL_ALIGN_BOTTOM;
var BLEND_SOURCE_OVER = 'srcOver';
exports.BLEND_SOURCE_OVER = BLEND_SOURCE_OVER;
var BLEND_DESTINATION_OVER = 'dstOver';
exports.BLEND_DESTINATION_OVER = BLEND_DESTINATION_OVER;
var BLEND_MULTIPLY = 'multiply';
exports.BLEND_MULTIPLY = BLEND_MULTIPLY;
var BLEND_SCREEN = 'screen';
exports.BLEND_SCREEN = BLEND_SCREEN;
var BLEND_OVERLAY = 'overlay';
exports.BLEND_OVERLAY = BLEND_OVERLAY;
var BLEND_DARKEN = 'darken';
exports.BLEND_DARKEN = BLEND_DARKEN;
var BLEND_LIGHTEN = 'lighten';
exports.BLEND_LIGHTEN = BLEND_LIGHTEN;
var BLEND_HARDLIGHT = 'hardLight';
exports.BLEND_HARDLIGHT = BLEND_HARDLIGHT;
var BLEND_DIFFERENCE = 'difference';
exports.BLEND_DIFFERENCE = BLEND_DIFFERENCE;
var BLEND_EXCLUSION = 'exclusion'; // Discover Jimp directory in any environment. (also in fucking karma)

exports.BLEND_EXCLUSION = BLEND_EXCLUSION;

function getJimpDir() {
  if ("production" === 'test' && undefined === 'browser') {
    return 'http://localhost:9876/base/test/';
  }

  return "".concat(__dirname, "/");
}

var dirName = getJimpDir(); // Font locations

exports.dirName = dirName;
var FONT_SANS_8_BLACK = dirName + '../fonts/open-sans/open-sans-8-black/open-sans-8-black.fnt';
exports.FONT_SANS_8_BLACK = FONT_SANS_8_BLACK;
var FONT_SANS_10_BLACK = dirName + '../fonts/open-sans/open-sans-10-black/open-sans-10-black.fnt';
exports.FONT_SANS_10_BLACK = FONT_SANS_10_BLACK;
var FONT_SANS_12_BLACK = dirName + '../fonts/open-sans/open-sans-12-black/open-sans-12-black.fnt';
exports.FONT_SANS_12_BLACK = FONT_SANS_12_BLACK;
var FONT_SANS_14_BLACK = dirName + '../fonts/open-sans/open-sans-14-black/open-sans-14-black.fnt';
exports.FONT_SANS_14_BLACK = FONT_SANS_14_BLACK;
var FONT_SANS_16_BLACK = dirName + '../fonts/open-sans/open-sans-16-black/open-sans-16-black.fnt';
exports.FONT_SANS_16_BLACK = FONT_SANS_16_BLACK;
var FONT_SANS_32_BLACK = dirName + '../fonts/open-sans/open-sans-32-black/open-sans-32-black.fnt';
exports.FONT_SANS_32_BLACK = FONT_SANS_32_BLACK;
var FONT_SANS_64_BLACK = dirName + '../fonts/open-sans/open-sans-64-black/open-sans-64-black.fnt';
exports.FONT_SANS_64_BLACK = FONT_SANS_64_BLACK;
var FONT_SANS_128_BLACK = dirName + '../fonts/open-sans/open-sans-128-black/open-sans-128-black.fnt';
exports.FONT_SANS_128_BLACK = FONT_SANS_128_BLACK;
var FONT_SANS_8_WHITE = dirName + '../fonts/open-sans/open-sans-8-white/open-sans-8-white.fnt';
exports.FONT_SANS_8_WHITE = FONT_SANS_8_WHITE;
var FONT_SANS_16_WHITE = dirName + '../fonts/open-sans/open-sans-16-white/open-sans-16-white.fnt';
exports.FONT_SANS_16_WHITE = FONT_SANS_16_WHITE;
var FONT_SANS_32_WHITE = dirName + '../fonts/open-sans/open-sans-32-white/open-sans-32-white.fnt';
exports.FONT_SANS_32_WHITE = FONT_SANS_32_WHITE;
var FONT_SANS_64_WHITE = dirName + '../fonts/open-sans/open-sans-64-white/open-sans-64-white.fnt';
exports.FONT_SANS_64_WHITE = FONT_SANS_64_WHITE;
var FONT_SANS_128_WHITE = dirName + '../fonts/open-sans/open-sans-128-white/open-sans-128-white.fnt'; // Edge Handling

exports.FONT_SANS_128_WHITE = FONT_SANS_128_WHITE;
var EDGE_EXTEND = 1;
exports.EDGE_EXTEND = EDGE_EXTEND;
var EDGE_WRAP = 2;
exports.EDGE_WRAP = EDGE_WRAP;
var EDGE_CROP = 3;
exports.EDGE_CROP = EDGE_CROP;
